torch/testing/_internal/common_utils.py

```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
+
+    def __init__(self, conn, test_case):
+        self.conn = conn
+        self.test_case = test_case
+        self.seen = set()
+
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        self.seen.add(func)
+        if kwargs is None:
+            kwargs = {}
+        return func(*args, **kwargs)
+
+    def commit(self):
+        rows = [
+            (
+                self.test_file,
+                self.test_case.id(),
+                torch.overrides.resolve_name(func) or str(func)
+            )
+            for func in self.seen
+        ]
+        self.conn.executemany("INSERT INTO calls_view VALUES (?,?,?)", rows)
+        self.conn.commit()
+
+    def __del__(self):
+        self.commit()
+
 class CrossRefMode(torch.overrides.TorchFunctionMode):
+    def __init__(self, test_case):
+        self.test_case = test_case
+
     def __torch_function__(self, func, types, args=(), kwargs=None):
         kwargs = kwargs or {}
+
+        hit = 0
+        miss = 0
+
+        # Doesn't actually return a storage
+        @functools.lru_cache(None)
+        def meta_storage(s):
+            return torch.empty(s.size(), dtype=s.dtype, device='meta')
+
+        def safe_is_leaf(t):
+            try:
+                return t.is_leaf
+            except RuntimeError:
+                # inference mode can trigger this
+                return False
+
+        @functools.lru_cache(None)
+        def meta_tensor(t):
+            with torch.inference_mode(t.is_inference()):
+                s = meta_storage(t.storage())
+                is_leaf = safe_is_leaf(t)
+                if is_leaf or not t._is_view():
+                    r = torch.empty(
+                        (0,), dtype=t.dtype, device='meta'
+                    )
+                    torch._C._set_storage_via_tensor(
+                        r, s, t.storage_offset(), t.size(), t.stride()
+                    )
+                    r.requires_grad = t.requires_grad
+                    if not is_leaf and t.requires_grad:
+                        with torch.enable_grad():
+                            r = r.clone()
+                else:
+                    base = torch.empty(
+                        (0,), dtype=t.dtype, device='meta'
+                    )
+                    torch._C._set_storage_via_tensor(
+                        base, s, 0, s.size(), (1,)
+                    )
+                    base.requires_grad = t.requires_grad
+                    with torch.enable_grad():
+                        if t._is_view() and not safe_is_leaf(t._base):
+                            base = base.clone()
+                        r = base.as_strided(t.size(), t.stride(), t.storage_offset())
+                torch._C._set_conj(r, t.is_conj())
+                torch._C._set_neg(r, t.is_neg())
+            return r
+
+        def to_meta(t):
+            nonlocal hit, miss
+            # TODO: zero tensors?  We appear to have eliminated them by
```

Comment:
Zero Tensors only appear in the forward AD for now (unless you do fun stuff). And all of that is in c++ so that sounds ok.
We could add a test with Custom Function that would trigger that though I think.